app/assets/javascripts/geoblacklight/viewers/wmts.js

```diff
@@ -0,0 +1,85 @@
+//= require geoblacklight/viewers/wms
+
+GeoBlacklight.Viewer.Wmts = GeoBlacklight.Viewer.Wms.extend({
+
+  addPreviewLayer: function() {
+    var _this = this;
+    fetch(this.data.url).then(function(response) {
+      if (!response.ok) {
+          throw new Error("Unable to fetch WMTSCapabilities document");
+      }
+      return response.text();
+    }).then(function(wmtsCapabilities) {
+      var tileInfo = _this.parseCapabilities(wmtsCapabilities),
+          options = tileInfo.bounds ? { bounds: tileInfo.bounds } : {},
+          tileLayer = L.tileLayer(tileInfo.url, options);
+      _this.overlay.addLayer(tileLayer);
+    }).catch(function(error) {
+      console.debug(error);
+    });
+  },
+
+  parseCapabilities: function(wmtsCapabilities) {
+    var parser = new DOMParser(),
+        doc = parser.parseFromString(wmtsCapabilities, "application/xml"),
+        layers = doc.getElementsByTagName("Layer"),
+        selectedLayer = this.findLayer(layers),
+        resourceElement = selectedLayer.getElementsByTagName("ResourceURL")[0];
+
+    // Generate URL with Leaflet.TileLayer URL template formatting
+    var tileURL = resourceElement.getAttribute("template")
+      .replace("{Style}", this.getStyle(selectedLayer))
+      .replace("{TileMatrixSet}", this.getTileMatrixSet(selectedLayer))
```

Comment:
are there any negative consequences if either of these are `undefined` (because there was no `<Style>` or `<TileMatrixSet>`)? can that happen?

Reply:
It's really up to the creator of the WMTSCapabilites document. `Style` and `TileMatrixSet` are required according to the spec, but sometimes the URL template doesn't make use of it. The [Princeton WMTSCapabilities](https://github.com/geoblacklight/geoblacklight/pull/1130/files#diff-18211e325f3cabced0e3dd77e290670f4d639591d364cd246e88482ad78d5ebdR51) fixture is an example. If the template requires them and they are aren't defined, then adding the layer will fail. Which is expected.

Reply:
Makes sense!